cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -71,6 +81,160 @@ public Schema getSchema() {
+                                                               new HashSet<>(
+      Arrays.asList(Schema.LogicalType.TIMESTAMP_MILLIS.name(), Schema.LogicalType.TIMESTAMP_MICROS.name())));
+
+    if (fields.get(fieldName) == null) {
+      return null;
+    }
+
+    if (logicalType == Schema.LogicalType.TIMESTAMP_MILLIS) {
+      return getZonedDateTime(fieldName, ZoneId.systemDefault(), TimeUnit.MILLISECONDS, 1000L);
+    }
+
+    return getZonedDateTime(fieldName, ZoneId.systemDefault(), TimeUnit.MICROSECONDS, 1000000L);
+  }
+
+  /**
+   * Get {@link ZonedDateTime} from field of logical type timestamp.
+   *
+   * @param fieldName zoned date time field to get.
+   * @param zoneId    zone id for the field
+   * @return value of the field as an object of {@link ZonedDateTime}
+   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
+   *                                  is not of {@link Schema.LogicalType#TIMESTAMP_MILLIS} or
+   *                                  {@link Schema.LogicalType#TIMESTAMP_MICROS} type.
+   */
+  public ZonedDateTime getTimestamp(String fieldName, ZoneId zoneId) {
+    Schema.LogicalType logicalType = validateAndGetLogicalType(fieldName, schema.getField(fieldName),
+                                                               new HashSet<>(
+      Arrays.asList(Schema.LogicalType.TIMESTAMP_MILLIS.name(), Schema.LogicalType.TIMESTAMP_MICROS.name())));
+
+    if (fields.get(fieldName) == null) {
+      return null;
+    }
+
+    if (logicalType == Schema.LogicalType.TIMESTAMP_MILLIS) {
+      return getZonedDateTime(fieldName, zoneId, TimeUnit.MILLISECONDS, 1000L);
+    }
+
+    return getZonedDateTime(fieldName, zoneId, TimeUnit.MICROSECONDS, 1000000L);
+  }
+
+  /**
+   * Get zoned date and time represented by the field.
+   *
+   * @param fieldName name of the field
+   * @param zoneId    zone id for the field
+   * @return {@link ZonedDateTime} represented by field.
+   */
+  private ZonedDateTime getZonedDateTime(String fieldName, ZoneId zoneId, TimeUnit unit, long divisor) {
+    Long ts = (Long) fields.get(fieldName);
+    // get fractional portion of timestamp, we always store timestamp in micro seconds, so first get that portion.
+    // for example: timestamp is 1970-01-01T00:00:01.112456Z, here fraction = 112456, tsInSeconds = 1
+    int fraction = (int) (ts % divisor);
+    long tsInSeconds = ts / divisor;
+    // create an Instant with time in seconds and fraction which will be stored as nano seconds.
+    Instant instant = Instant.ofEpochSecond(tsInSeconds, unit.toNanos(fraction));
+    return ZonedDateTime.ofInstant(instant, zoneId);
+  }
+
+  private static Schema.LogicalType validateAndGetLogicalType(String fieldName, Schema.Field field,
+                                                              Set<String> logicalTypes)
+    throws IllegalArgumentException {
```

Comment:
Usually for runtime exception, you don't declare it in the signature but rather javadoc'ed.